File: prettier-plugin-csharpier/src/Printer/Types/UsingStatement.ts
import { Doc, FastPath, ParserOptions } from "prettier";
import { Print, PrintMethod } from "../PrintMethod";
import { SyntaxToken, printPathSyntaxToken, SyntaxTreeNode } from "../SyntaxTreeNode";
import { concat, group, hardline, indent, join, softline, line, doubleHardline } from "../Builders";
import { AttributeListNode } from "./AttributeList";
import { printVariableDeclaration, VariableDeclarationNode } from "./VariableDeclaration";

export interface UsingStatementNode extends SyntaxTreeNode<"UsingStatement"> {
    attributeLists: AttributeListNode[];
    awaitKeyword?: SyntaxToken;
    usingKeyword?: SyntaxToken;
    openParenToken?: SyntaxToken;
    declaration?: VariableDeclarationNode;
    expression?: SyntaxTreeNode;
    closeParenToken?: SyntaxToken;
    statement?: SyntaxTreeNode;
}

export const printUsingStatement: PrintMethod<UsingStatementNode> = (path, options, print) => {
    const node = path.getValue();
    const parts: Doc[] = [
        printPathSyntaxToken(path, "usingKeyword"),
        " (",
        node.declaration ? path.call(o => printVariableDeclaration(o, options, print), "declaration") : "",
        node.expression ? path.call(print, "expression") : "",
        ")"
    ];

    let statement = path.call(print, "statement");
    if (node.statement?.nodeType === "UsingStatement") {
        parts.push(hardline, statement);
    } else if (node.statement?.nodeType === "Block") {
        parts.push(statement);
    } else {
        parts.push(indent(concat([hardline, statement])));
    }

    return concat(parts);
};

File: Src/CSharpier/SyntaxPrinter/SyntaxNodePrinters/UsingStatement.cs
namespace CSharpier.SyntaxPrinter.SyntaxNodePrinters;

internal static class UsingStatement
{
    public static Doc Print(UsingStatementSyntax node, PrintingContext context)
    {
        var docs = new ValueListBuilder<Doc>([null, null, null, null]);
        docs.Append(ExtraNewLines.Print(node));
        docs.Append(
            Doc.Group(
                Token.Print(node.AwaitKeyword, context),
                node.AwaitKeyword.RawSyntaxKind() != SyntaxKind.None ? " " : Doc.Null,
                Token.Print(node.UsingKeyword, context),
                " ",
                Token.Print(node.OpenParenToken, context),
                Doc.Group(
                    Doc.Indent(
                        Doc.SoftLine,
                        node.Declaration != null
                            ? VariableDeclaration.Print(node.Declaration, context)
                            : Doc.Null,
                        node.Expression != null ? Node.Print(node.Expression, context) : Doc.Null
                    ),
                    Doc.SoftLine
                ),
                Token.Print(node.CloseParenToken, context),
                Doc.IfBreak(Doc.Null, Doc.SoftLine)
            )
        );

        if (node.Statement is UsingStatementSyntax)
        {
            docs.Append(Doc.HardLine, Node.Print(node.Statement, context));
        }
        else if (node.Statement is BlockSyntax blockSyntax)
        {
            docs.Append(Block.Print(blockSyntax, context));
        }
        else
        {
            docs.Append(Doc.Indent(Doc.HardLine, Node.Print(node.Statement, context)));
        }

        return Doc.Concat(ref docs);
    }
}
